### PR TITLE
Remove `Session::one_time_diagnostic`

### PR DESCRIPTION
**Pull request by cjgillot**
This is untracked mutable state, which modified the behaviour of queries.
It was used for 2 things: some full-blown errors, but mostly for lint declaration notes ("the lint level is defined here" notes).

It is replaced by the diagnostic deduplication infra which already exists in the diagnostic emitter.
A new diagnostic level `OnceNote` is introduced specifically for lint notes, to deduplicate subdiagnostics.

As a drive-by, diagnostic emission takes a `&mut` to allow dropping the `SubDiagnostic`s.

### CODE REVIEW

File: compiler/rustc_errors/src/lib.rs [hunk 1]
```diff
@@ -993,25 +994,25 @@ impl HandlerInner {
...
-            self.emit_diagnostic(diag);
-        });
+            self.emit_diagnostic(&mut diag);
+        }
         reported
     }
 
     // FIXME(eddyb) this should ideally take `diagnostic` by value.
-    fn emit_diagnostic(&mut self, diagnostic: &Diagnostic) -> Option<ErrorGuaranteed> {
+    fn emit_diagnostic(&mut self, diagnostic: &mut Diagnostic) -> Option<ErrorGuaranteed> {
```
Comment (estebank) [inline compiler/rustc_errors/src/lib.rs:1015]: Would fixing the `FIXME` be possible as part of this PR, or are there big consequences that I'm failing to think of?

File: src/test/ui-fulldeps/lint-tool-test.stderr [hunk 2]
```diff
@@ -83,11 +83,6 @@ error: item is named 'lintmetoo'
 LL |     fn lintmetoo() { }
    |     ^^^^^^^^^^^^^^^^^^
    |
-note: the lint level is defined here
-  --> $DIR/lint-tool-test.rs:14:9
-   |
-LL | #![deny(clippy_group)]
-   |         ^^^^^^^^^^^^
```
[the remark this answers is not in the document]
Comment (estebank) [reply src/test/ui-fulldeps/lint-tool-test.stderr:90]: Oh! Wait, the note is still being emitted, but now we don't point at the group. That's perfectly fine.

**Review by estebank — APPROVED** [review]
I'm r=me on the changes, but the tests are failing.

CC @eddyb, who was touching this code recently just in case (feel free to ignore).

File: compiler/rustc_errors/src/lib.rs (hunk 1 above)
Comment (cjgillot) [reply compiler/rustc_errors/src/lib.rs:1015]: This requires to make the `DiagnosticBuilder::emit` method take `self` by move, which breaks `struct_err(..).label(..).emit()` chains.

**Comment by cjgillot** [comment]
rustdoc-ui tests have been blessed, with the expected behaviour.
@bors r=estebank

File: compiler/rustc_errors/src/lib.rs (hunk 1 above)
Comment (eddyb) [reply compiler/rustc_errors/src/lib.rs:1015]: Not exactly, I decided we can't make `.emit()` take `&mut self` for that reason, but once `.emit()` is called the `DiagnosticBuilder` state changes to ignore the `Diagnostic` on further `.emit()`, so we could be replacing the `Diagnostic` with a dummy one (or even having `Option` around it but that could slow down the `Deref`/`DerefMut` impl a bunch).

**Pull request merged**